stm32h7rs run application in external flash memory XIP

FRASTM opened the pull request:
the stm32h7rs nucleo or disco kit boards can run an application in external flash
The internal flash memory is limited to 64KB which is enough for mcuboot but might be too small for some zephyr applications

With this PR it enable the memorymapped mode on the external flash memory so that XIP is possible on the external NOR octoflash

Two samples are running in XIP on the external flash 
- samples/application_development/code_relocation_nocopy/
- samples/boards/st/hello_world_xip/ built with MCUboot https://github.com/mcu-tools/mcuboot/pull/2323



This PR completes the https://github.com/zephyrproject-rtos/zephyr/pull/88051
with samples to run application in external NOR memory on stm32h7rs target boards

FRASTM committed:
boards: arm: stm32h7s78 disco kit for XiP on external flash

Add the memory attributes to XiP in MemoryMapped mode on the
external octo-flash of the stm32h7s78_dk board.
Use the STM32Cube programmer to flash with the corresponding
external loader for XiP mode.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: st: stm32h7s3l8 nucleo board for XiP on external flash

Add the memory attributes to XiP in MemoryMapped mode on the
external octo-flash of the nucleo_h7s3l8 board.
Use the STM32Cube programmer to flash with the corresponding
external loader for XiP mode.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Run the samples/application_development/code_relocation_nocopy/

```
*** Booting Zephyr OS build v4.1.0-1733-g59dd3dd8abac ***                                                                                      
Address of main function 0x8000685                                                                                                             
Address of function_in_ext_flash 0x70000001                                                                                                    
Address of var_ext_sram_data 0x240000e0 (10)                                                                                                   
Address of function_in_sram 0x24000001                                                                                                         
Address of var_sram_data 0x240000e4 (10)                                                                                                       
Hello World! nucleo_h7s3l8
```

FRASTM commented:
With the samples/boards/st/hello_world_xip/ the goal is to place the mcuboot in the internalm flash memory and to jump to a zephyr application executed in  the external flash
This is possible with the mcuboot as in https://github.com/mcu-tools/mcuboot/pull/2154
plus the application built/linkedfor external memory  ` west build -p -b nucleo_h7s3l8 samples/boards/st/hello_world_xip/ --sysbuild -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y`
then flashing the

titouanc commented on boards/st/stm32h7s78_dk/stm32h7s78_dk.dts:
```diff
@@ -173,6 +175,64 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			slot0_partition: partition@0 {
+				label = "image-0";
+				reg = <0x00000000 DT_SIZE_K(416)>;
+			};
+			slot1_partition: partition@80000 {
+				label = "image-1";
+				reg = <0x00080000 DT_SIZE_K(416)>;
+			};
+			scratch_partition: partition@e0000 {
+				label = "image-scratch";
+				reg = <0x000e0000 DT_SIZE_K(64)>;
+			};
+			storage_partition: partition@f0000 {
+				label = "storage";
+				reg = <0x000f0000 DT_SIZE_K(64)>;
+			};
```
Wrong offsets (0x00080000 is not 416k)

FRASTM replied:
Yes, using the same partitioning as the nucleo_h7s3l8

titouanc commented on dts/arm/st/h7rs/stm32h7rs.dtsi:
```diff
@@ -81,7 +82,8 @@
 		reg = <0x70000000 DT_SIZE_M(256)>;
 		zephyr,memory-region = "EXTMEM";
-		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_EXTMEM) )>;
+		/* The ATTR_MPU_EXTMEM attribut causing a MPU FAULT */
```
typo

```suggestion
		/* The ATTR_MPU_EXTMEM attribute causing a MPU FAULT */
```

titouanc commented on samples/boards/st/hello_world_xip/README.rst:
```diff
@@ -0,0 +1,61 @@
+********************
+
+This application can be built with 
+west build -p always -b nucleo_h7s3l8 samples/boards/stm32/hello_world_xip/ --sysbuild -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y
```
```suggestion
west build -p always -b nucleo_h7s3l8 samples/boards/st/hello_world_xip/ --sysbuild -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y
```

FRASTM replied:
fixed

titouanc commented:
Hello @FRASTM I just tested this PR with the STM32H7S78-DK. First of all, I had to make these changes to have something that builds & flash properly:

```diff
diff --git a/samples/boards/st/hello_world_xip/boards/stm32h7s78_dk.overlay b/samples/boards/st/hello_world_xip/boards/stm32h7s78_dk.overlay
index 700cb5f56d5..5357426313d 100644
--- a/samples/boards/st/hello_world_xip/boards/stm32h7s78_dk.overlay
+++ b/samples/boards/st/hello_world_xip/boards/stm32h7s78_dk.overlay
@@ -13,6 +13,11 @@
                zephyr,flash = &mx66uw1g45;
                zephyr,code-partition = &slot0_partition;
        };
+
+       aliases {
+               led1 = &green_led;
+               led2 = &orange_led;
+       };
 };
 
 &mx66uw1g45 {
```

I am building with the following command:
```bash
west build -b stm32h7s78_dk zephyr/samples/boards/st/hello_world_xip/ --sysbuild -- \
    -DSB_CONFIG_BOOTLOADER_MCUBOOT=y \
    -Dhello_world_xip_CONFIG_FLASH=y -Dhello_world_xip_CONFIG_STM32_MEMMAP=y \
    -Dmcuboot_CONFIG_LOG=y -Dmcuboot_CONFIG_LOG_DEFAULT_LEVEL=3
```

I am getting this log on the console: 
```
E: Could not select XSPI domain clock
*** Booting MCUboot v2.1.0-269-gc8470fb145f8 ***
*** Using Zephyr OS build v4.1.0-1736-gbcc9c4ce39d8 ***
I: Starting bootloader
E: Failed to open flash area ID 1 (image 0 slot 0): -19, cannot continue
```

From what I understand, the message `E: Could not select XSPI domain clock` is coming from the xspi flash driver that attempts to enable the clock source PLL2_S; but PLL2 is disabled in the device tree.

titouanc commented:
I've added the following pll2 configuration (same as the pll1 above):

```diff
diff --git a/boards/st/stm32h7s78_dk/stm32h7s78_dk.dts b/boards/st/stm32h7s78_dk/stm32h7s78_dk.dts
index d42b89d1629..b51d53e2ed2 100644
--- a/boards/st/stm32h7s78_dk/stm32h7s78_dk.dts
+++ b/boards/st/stm32h7s78_dk/stm32h7s78_dk.dts
@@ -88,6 +90,19 @@
        status = "okay";
 };
 
+/* PLL2 for clocking the xspi peripheral */
+&pll2 {
+       clocks = <&clk_hse>;
+       div-m = <12>;
+       mul-n = <250>;
+       div-p = <2>;
+       div-q = <2>;
+       div-r = <2>;
+       div-s = <2>;
+       div-t = <2>;
+       status = "okay";
+};
+
 &rcc {
        clocks = <&pll>;
        clock-frequency = <DT_FREQ_M(250)>;

```

I don't know if that's a valid configuration at all, but now I am getting this log:

```
I: XSPI flash config is OCTO / DTR
I: Read SFDP from externalFlash
I: Read SFDP from externalFlash
I: NOR external-flash at 0x70000000 (0x8000000 bytes)
*** Booting MCUboot v2.1.0-269-gc8470fb145f8 ***
*** Using Zephyr OS build v4.1.0-1736-gbcc9c4ce39d8 ***
I: Starting bootloader
E: 2: Failed to send XSPI instruction
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x0, image_ok=0x24
E: 2: Failed to send XSPI instruction
I: Secondary image: magic=bad, swap_type=0xf, copy_done=0x0, image_ok=0x24
I: Boot source: none
E: 2: Failed to send XSPI instruction
W: Failed reading image headers; Image=0
E: 2: Failed to send XSPI instruction
E: Image in the primary slot is not valid!
E: Unable to find bootable image
```

FRASTM commented on samples/boards/st/hello_world_xip/prj.conf:
Not here, the xspi flash is configured in MemMapped from the mcuboot point of view
There is no need for configuring the MEMMAP from this hello_world_xip application POV : this application is not supposed to access the xspi for its own purpose, the external flash is "reserved" for executing the application

FRASTM replied:
Up to now I download with stm32cubeprogrammer GUI 
the build/mcuboot/zephyr/zephyr.bin at internal flash address 0x8000000
the build/hello_world_xip/zephyr/zephyr.signed.bin at external flash address 0x70000000 (or 0x90000000)

FRASTM commented on samples/boards/st/hello_world_xip/boards/nucleo_h7s3l8.overlay:
```diff
@@ -0,0 +1,41 @@
+};
+
+&mx25uw25645 {
+		partitions {
```
yes,

FRASTM commented on samples/boards/st/hello_world_xip/src/main.c:
```diff
@@ -0,0 +1,88 @@
+
+#define __ASM __asm /*!< asm keyword for GNU Compiler */
+#define __INLINE inline /*!< inline keyword for GNU Compiler */
+#define __STATIC_INLINE static inline
```
removed

FRASTM commented on samples/boards/st/hello_world_xip/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+      - b_u585i_iot02a
+    platform_allow:
+      - nucleo_h7s3l8
+      - stm32h7s78_dk
```
yes, as soon as this is tested, 
--> coming soon

FRASTM replied:
The stm32h750b_dk has quadspi NOR, that will be handled by another PR

FRASTM commented on samples/boards/st/hello_world_xip/src/main.c:
```diff
@@ -0,0 +1,88 @@
+#define __INLINE inline /*!< inline keyword for GNU Compiler */
+#define __STATIC_INLINE static inline
+
+__attribute__((always_inline)) __STATIC_INLINE uint32_t __get_PC(void)
```
ok

FRASTM commented on boards/st/nucleo_h7s3l8/nucleo_h7s3l8.dts:
```diff
@@ -20,6 +20,7 @@
 		zephyr,dtcm = &dtcm;
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
+		zephyr,flash-controller = &mx25uw25645;
```
Ok, the `zephyr,flash-controller = &mx25uw25645` is for the overlay pf the sample

FRASTM commented on dts/arm/st/h7rs/stm32h7rs.dtsi:
```diff
@@ -769,6 +771,32 @@
 
+		xspi1: spi@52005000 {
+			compatible = "st,stm32-xspi";
+			reg = <0x52005000 0x1000>;
```
From the ref man  [RM0477 STM32H7Rx/7Sx Arm®-based 32-bit MCUs ](https://www.st.com/resource/en/reference_manual/rm0477-stm32h7rx7sx-armbased-32bit-mcus-stmicroelectronics.pdf) 

0x52005000 - 0x52005FFF XSPI1

FRASTM commented on dts/arm/st/h7rs/stm32h7rs.dtsi:
```diff
@@ -769,6 +771,32 @@
+
+		xspi2: spi@5200a000 {
+			compatible = "st,stm32-xspi";
+			reg = <0x5200a000 0x1000>;
```
From the ref man  [RM0477 STM32H7Rx/7Sx Arm®-based 32-bit MCUs ](https://www.st.com/resource/en/reference_manual/rm0477-stm32h7rx7sx-armbased-32bit-mcus-stmicroelectronics.pdf) 

0x5200A000 - 0x5200AFFF XSPI2

FRASTM commented:
CI failure is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/88646PR

FRASTM commented on boards/st/nucleo_h7s3l8/nucleo_h7s3l8.dts:
```diff
@@ -53,6 +53,7 @@
 		led2 = &red_led;
 		sw0 = &user_button;
 		watchdog0 = &iwdg;
+		spi-flash0 = &mx25uw25645;
```
in the https://github.com/mcu-tools/mcuboot/blob/main/boot/zephyr/flash_map_extended.c, the FLASH_DEVICE_ID is defined as SPI_FLASH_0_ID

nordicjm replied:
it is not, it is like that for specific platforms that the owners have added

FRASTM commented:
Based on the https://github.com/zephyrproject-rtos/zephyr/pull/88579

FRASTM committed:
samples: appli dev: code reloc running on stm32h7rs target boards

Samples to execute code in external memory in Memory mapped mode
on the stm32h7s3l8 nucleo or stm32h7s78 disco kit

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
set the DNM label until the https://github.com/zephyrproject-rtos/zephyr/pull/88051 is merged

FRASTM committed:
samples: sysbuild: with mcu_boot on the stm32h7rs targets

Run the sample to execute in place on the external flash of the
stm32h7s78_dk or nucleo_h7s3l8 boards.
with_boot is build and linked for the external flash
Download the mcuboot zephyr.bin in internal memory
Download the with_mcuboot zephyr.signed.bin in external memory

Signed-off-by: Francois Ramu <francois.ramu@st.com>

JarmouniA commented:
> When the samples/application_development/code_relocation_nocopy/ is built, the CONFIG_BOOTLOADER_MCUBOOT is not defined, but CONFIG_STM32_MEMMAP is. Should both be kept ?
> 
> ```
> if(CONFIG_STM32_MEMMAP OR (CONFIG_XIP AND CONFIG_BOOTLOADER_MCUBOOT))
> board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
> board_runner_args(stm32cubeprogrammer "--extload=MX25UW25645G_NUCLEO-H7S3L8.stldr")
> else()
> ...
> ```

https://github.com/zephyrproject-rtos/zephyr/pull/88790#discussion_r2077541765

titouanc commented:
Hello @FRASTM I just tested this PR on my STM32H7S78-DK. Invoking west like this:

```bash
west build -b stm32h7s78_dk samples/sysbuild/with_mcuboot/ --sysbuild -- \
     -Dmcuboot_CONFIG_FLASH_LOG_LEVEL_DBG=y
```

However, it still doesn't boot the image on the external flash, see the log below:

<details>

<summary>Output from the board's console</summary>

```
D: Flash initialized. BS: 32
D: Block 0: bs: 8192 count: 8
D: XSPI Init'd
D: Flash initialized. BS: 32
D: Block 0: bs: 8192 count: 8
D: XSPI Init'd
D: Reset Mem (SPI/STR)
D: Status Match cb
D: Mem Ready (SPI/STR)
D: XSPI configuring Octo SPI mode
D: Status Match cb
D: Status Match cb
D: Status Match cb
D: Status Match cb
I: XSPI flash config is OCTO / DTR
I: Read SFDP from externalFlash
D: xspi-nor-flash@0: SFDP v 1.8 AP fd with 5 PH
D: PH0: ff00 rev 1.7: 20 DW @ 40
I: Read SFDP from externalFlash
D: xspi-nor-flash@0: 128 MiBy flash
D: Erase 4096 with 20
D: Erase 65536 with d8
D: Address width: 4 Bytes
D: Page size 256 bytes
D: Flash size 134217728 bytes
D: Using read mode: 0, instr: 0x0, dummy cycles: 0
D: Using write instr: 0x12ED
D: PH1: ff87 rev 1.1: 28 DW @ 90
D: layout page 256 not compatible with erase size 4096
D: erase size will be used as layout page size
D: layout 32768 x 4096 By pages
I: NOR external-flash at 0x70000000 (0x8000000 bytes)
*** Booting MCUboot v2.1.0-rc1-310-g990b1fcb367e ***
*** Using Zephyr OS build v4.1.0-5079-g2506a1fea8f9 ***
D: XSPI: read 32 data at 0x70000000
D: Instruction 0xee11
D: Rx Cplt cb
D: XSPI: read 32 data at 0x70080000
D: Instruction 0xee11
D: Rx Cplt cb
D: XSPI: read 16 data at 0x7007fff0
D: Instruction 0xee11
D: Rx Cplt cb
D: XSPI: read 1 data at 0x7007ff80
D: Instruction 0xee11
D: Rx Cplt cb
D: XSPI: read 1 data at 0x7007ffa0
D: Instruction 0xee11
E: 2: Failed to send XSPI instruction
D: XSPI: read 16 data at 0x700ffff0
D: Instruction 0xee11
E: 2: Failed to send XSPI instruction
D: XSPI: read 32 data at 0x70000000
D: Instruction 0xee11
E: 2: Failed to send XSPI instruction
W: Failed reading image headers; Image=0
D: XSPI: read 256 data at 0x70000000
D: Instruction 0xee11
E: 2: Failed to send XSPI instruction
E: Image in the primary slot is not valid!
E: Unable to find bootable image
```
</details>

Also, each time the message `E: 2: Failed to send XSPI instruction` appears, there is a delay of a few seconds; so it smells like a timeout condition in the xspi flash driver (here I think: https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32h7rsxx/drivers/src/stm32h7rsxx_hal_xspi.c#L3157)

nordicjm commented on boards/st/nucleo_h7s3l8/board.cmake:
```diff
@@ -1,7 +1,12 @@
 # keep first
+if(CONFIG_STM32_MEMMAP OR (CONFIG_XIP AND CONFIG_BOOTLOADER_MCUBOOT))
 board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
+board_runner_args(stm32cubeprogrammer "--extload=MX25UW25645G_NUCLEO-H7S3L8.stldr")
```
2 space indent for cmake

FRASTM replied:
done

nordicjm commented on boards/st/stm32h7s78_dk/board.cmake:
```diff
@@ -1,7 +1,13 @@
 # keep first
+if(CONFIG_STM32_MEMMAP OR (CONFIG_XIP AND CONFIG_BOOTLOADER_MCUBOOT))
 board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
+board_runner_args(stm32cubeprogrammer "--extload=MX66UW1G45G_STM32H7S78-DK.stldr")
```
as above

FRASTM replied:
done

FRASTM commented:
> Hello @FRASTM I just tested this PR on my STM32H7S78-DK. Invoking west like this:
> 
> ```shell
> west build -b stm32h7s78_dk samples/sysbuild/with_mcuboot/ --sysbuild -- \
>      -Dmcuboot_CONFIG_FLASH_LOG_LEVEL_DBG=y \
>      -Dmcuboot_CONFIG_LOG_MODE_IMMEDIATE=y
> ```
> 
> However, it still doesn't boot the image on the external flash, see the log below:
> 
> Output from the board's console
> Also, each time the message `<err> flash_stm32_xspi: 2: Failed to send XSPI instruction` appears, there is a delay of a few seconds; so it smells like a timeout condition in the xspi flash driver (here I think: https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32h7rsxx/drivers/src/stm32h7rsxx_hal_xspi.c#L3157)

Hello,
The log shows th

> Hello @FRASTM I just tested this PR on my STM32H7S78-DK. Invoking west like this:
> 
> ```shell
> west build -b stm32h7s78_dk samples/sysbuild/with_mcuboot/ --sysbuild -- \
>      -Dmcuboot_CONFIG_FLASH_LOG_LEVEL_DBG=y \
>      -Dmcuboot_CONFIG_LOG_MODE_IMMEDIATE=y
> ```
> 
> However, it still doesn't boot the image on the external flash, see the log below:
> 
> Output from the board's console
> Also, each time the message `<err> flash_stm32_xspi: 2: Failed to send XSPI instruction` appears, there is a delay of a few seconds; so it smells like a timeout condition in the xspi flash driver (here I think: https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32h7rsxx/drivers/src/stm32h7rsxx_hal_xspi.c#L3157)

Hello,
From the log, the external octo-NOR is accessed in indirect mode for reading and not in memory mapped.
Did you take the https://github.com/mcu-tools/mcuboot/pull/2323 ?

etienne-lms commented on boards/st/nucleo_h7s3l8/board.cmake:
```diff
@@ -1,7 +1,12 @@
+if(CONFIG_STM32_MEMMAP OR (CONFIG_XIP AND CONFIG_BOOTLOADER_MCUBOOT))
+  board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
+  board_runner_args(stm32cubeprogrammer "--extload=MX25UW25645G_NUCLEO-H7S3L8.stldr")
+else()
```
Maybe simplify to:
```patch
 board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
+if(...)
+  board_runner_args(stm32cubeprogrammer "--extload=MX25UW25645G_NUCLEO-H7S3L8.stldr")
+endif()

titouanc commented:
Hello @FRASTM !

> From the log, the external octo-NOR is accessed in indirect mode for reading and not in memory mapped.
> Did you take the mcu-tools/mcuboot#2323 ?

Indeed, it got lost in the PR discussion and I had forgotten about it. I just checked out your MCUboot PR (in bootloader/mcuboot), and tested again with the same west command as above. I don't see any error message from MCUBoot anymore, but the application is still not booting. In the log below, we can see that many more reads from flash succeed, but I don't see the Zephyr banner showing up when the application is supposed to boot.

<details>
<summary>Output from the board's console</summary>
[00:00:00.000,000] <dbg> flash_stm32h7: stm32h7_flash_init: Flash initialized. BS: 32
[00:00:00.008,000] <dbg> flash_stm32h7: stm32h7_flash_init: Block 0: bs: 8192 count: 8
[00:00:00.016,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_init: XSPI Init'd
[00:00:00.000,000] <dbg> flash_stm32h7: stm32h7_flash_init: Flash initialized. BS: 32
[00:00:00.008,000] <dbg> flash_stm32h7: stm32h7_flash_init: Block 0: bs: 8192 count: 8
[00:00:00.016,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_init: XSPI Init'd
[00:00:00.124,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_init: Reset Mem (SPI/STR)
[00:00:00.132,000] <dbg> flash_stm32_xspi: HAL_XSPI_StatusMatchCallback: Status Match cb
[00:00:00.140,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_init: Mem Ready (SPI/STR)
[00:00:00.148,000] <dbg> flash_stm32_xspi: stm32_xspi_config_mem: XSPI configuring Octo SPI mode
[00:00:00.158,000] <dbg> flash_stm32_xspi: HAL_XSPI_StatusMatchCallback: Status Match cb
[00:00:00.166,000] <dbg> flash_stm32_xspi: HAL_XSPI_StatusMatchCallback: Status Match cb
[00:00:00.175,000] <dbg> flash_stm32_xspi: HAL_XSPI_StatusMatchCallback: Status Match cb
[00:00:00.223,000] <dbg> flash_stm32_xspi: HAL_XSPI_StatusMatchCallback: Status Match cb
[00:00:00.232,000] <inf> flash_stm32_xspi: XSPI flash config is OCTO / DTR
[00:00:00.239,000] <inf> flash_stm32_xspi: Read SFDP from externalFlash
[00:00:00.246,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_init: xspi-nor-flash@0: SFDP v 1.8 AP fd with 5 PH
[00:00:00.256,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_init: PH0: ff00 rev 1.7: 20 DW @ 40
[00:00:00.266,000] <inf> flash_stm32_xspi: Read SFDP from externalFlash
[00:00:00.273,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: xspi-nor-flash@0: 128 MiBy flash
[00:00:00.282,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: Erase 4096 with 20
[00:00:00.290,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: Erase 65536 with d8
[00:00:00.298,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: Address width: 4 Bytes
[00:00:00.306,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: Page size 256 bytes
[00:00:00.314,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: Flash size 134217728 bytes
[00:00:00.323,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: Using read mode: 0, instr: 0x0, dummy cycles: 0
[00:00:00.333,000] <dbg> flash_stm32_xspi: spi_nor_process_bfp: Using write instr: 0x12ED
[00:00:00.342,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_init: PH1: ff87 rev 1.1: 28 DW @ 90
[00:00:00.351,000] <dbg> flash_stm32_xspi: setup_pages_layout: layout page 256 not compatible with erase size 4096
[00:00:00.362,000] <dbg> flash_stm32_xspi: setup_pages_layout: erase size will be used as layout page size
[00:00:00.372,000] <dbg> flash_stm32_xspi: setup_pages_layout: layout 32768 x 4096 By pages
[00:00:00.381,000] <dbg> flash_stm32_xspi: stm32_xspi_set_memorymap: MemoryMap mode enabled
[00:00:00.389,000] <inf> flash_stm32_xspi: Memory-mapped NOR-flash at 0x70000000 (0x8000000 bytes)
*** Booting MCUboot v2.1.0-rc1-315-ge3b8bd449b18 ***
*** Using Zephyr OS build v4.1.0-5079-g0ca62307b648 ***
[00:00:00.409,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000000, len 32
[00:00:00.419,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70080000, len 32
[00:00:00.429,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007fff0, len 16
[00:00:00.439,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007ff80, len 1
[00:00:00.450,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007ffa0, len 1
[00:00:00.460,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007ffc0, len 1
[00:00:00.470,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700ffff0, len 16
[00:00:00.480,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700fff80, len 1
[00:00:00.490,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700fffa0, len 1
[00:00:00.500,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700fffc0, len 1
[00:00:00.511,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000000, len 32
[00:00:00.521,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70080000, len 32
[00:00:00.531,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007fff0, len 16
[00:00:00.541,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007ff80, len 1
[00:00:00.551,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007ffa0, len 1
[00:00:00.562,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x7007ffc0, len 1
[00:00:00.572,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700ffff0, len 16
[00:00:00.582,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700fff80, len 1
[00:00:00.592,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700fffa0, len 1
[00:00:00.602,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x700fffc0, len 1
[00:00:00.612,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000000, len 256
[00:00:00.623,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000100, len 256
[00:00:00.634,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000200, len 256
[00:00:00.644,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000300, len 256
[00:00:00.655,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000400, len 256
[00:00:00.665,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000500, len 256
[00:00:00.676,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000600, len 256
[00:00:00.686,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000700, len 256
[00:00:00.697,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000800, len 256
[00:00:00.707,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000900, len 256
[00:00:00.718,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000a00, len 256
[00:00:00.728,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000b00, len 256
[00:00:00.739,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000c00, len 256
[00:00:00.749,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000d00, len 256
[00:00:00.760,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000e00, len 256
[00:00:00.770,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70000f00, len 256
[00:00:00.781,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001000, len 256
[00:00:00.791,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001100, len 256
[00:00:00.802,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001200, len 256
[00:00:00.812,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001300, len 256
[00:00:00.823,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001400, len 256
[00:00:00.834,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001500, len 256
[00:00:00.844,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001600, len 256
[00:00:00.855,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001700, len 256
[00:00:00.865,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001800, len 256
[00:00:00.876,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001900, len 256
[00:00:00.886,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001a00, len 256
[00:00:00.897,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001b00, len 256
[00:00:00.907,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001c00, len 256
[00:00:00.918,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001d00, len 256
[00:00:00.928,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001e00, len 256
[00:00:00.939,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70001f00, len 256
[00:00:00.949,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002000, len 256
[00:00:00.960,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002100, len 256
[00:00:00.970,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002200, len 256
[00:00:00.981,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002300, len 256
[00:00:00.991,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002400, len 256
[00:00:01.002,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002500, len 256
[00:00:01.012,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002600, len 256
[00:00:01.023,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002700, len 256
[00:00:01.033,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002800, len 256
[00:00:01.044,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002900, len 256
[00:00:01.054,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002a00, len 256
[00:00:01.065,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002b00, len 256
[00:00:01.075,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002c00, len 256
[00:00:01.086,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002d00, len 256
[00:00:01.097,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002e00, len 256
[00:00:01.107,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70002f00, len 256
[00:00:01.118,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003000, len 256
[00:00:01.128,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003100, len 256
[00:00:01.139,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003200, len 256
[00:00:01.149,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003300, len 256
[00:00:01.160,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003400, len 256
[00:00:01.170,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003500, len 256
[00:00:01.181,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003600, len 256
[00:00:01.191,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003700, len 256
[00:00:01.202,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003800, len 256
[00:00:01.212,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003900, len 256
[00:00:01.223,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003a00, len 256
[00:00:01.233,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003b00, len 256
[00:00:01.244,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003c00, len 256
[00:00:01.254,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003d00, len 256
[00:00:01.265,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003e00, len 256
[00:00:01.275,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70003f00, len 256
[00:00:01.286,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004000, len 256
[00:00:01.296,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004100, len 256
[00:00:01.307,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004200, len 256
[00:00:01.317,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004300, len 256
[00:00:01.328,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004400, len 256
[00:00:01.339,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004500, len 256
[00:00:01.349,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004600, len 256
[00:00:01.360,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004700, len 256
[00:00:01.370,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004800, len 256
[00:00:01.381,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004900, len 256
[00:00:01.391,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004a00, len 108
[00:00:01.401,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004a6c, len 4
[00:00:01.412,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004a70, len 4
[00:00:01.422,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004a74, len 32
[00:00:01.432,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004a94, len 4
[00:00:01.442,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004a98, len 32
[00:00:01.453,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004ab8, len 4
[00:00:01.463,000] <dbg> flash_stm32_xspi: flash_stm32_xspi_read: Memory-mapped read from 0x70004abc, len 256
</details>

Furthermore, I've modified the application to add a blinky like this (in case the application booted correctly but there's an UART issue), but I don't see the LED blinnking so it looks to me that the application hasn't booted at all.

```diff
diff --git a/samples/sysbuild/with_mcuboot/src/main.c b/samples/sysbuild/with_mcuboot/src/main.c
index 6a6197a801b..ee11a2d9de8 100644
--- a/samples/sysbuild/with_mcuboot/src/main.c
+++ b/samples/sysbuild/with_mcuboot/src/main.c
@@ -7,9 +7,21 @@
 #include <zephyr/kernel.h>
 #include <zephyr/linker/linker-defs.h>
 
+#include <zephyr/drivers/gpio.h>
+
+static const struct gpio_dt_spec led = GPIO_DT_SPEC_GET(DT_ALIAS(led0), gpios);
+
 int main(void)
 {
        printk("Address of sample %p\n", (void *)__rom_region_start);
        printk("Hello sysbuild with mcuboot! %s\n", CONFIG_BOARD);
+
+       gpio_pin_configure_dt(&led, GPIO_OUTPUT_INACTIVE);
+
+       while (1) {
+               gpio_pin_toggle_dt(&led);
+               k_msleep(500);
+       }
+
        return 0;
 }
```

FRASTM commented:
I confirm that  `west build -p -b nucleo_h7s3l8 samples/sysbuild/with_mcuboot/ --sysbuild`
makes the sample  running correctly and led0 is blinking

```
*** Booting MCUboot v2.1.0-rc1-316-g7865eeb124cf ***
*** Using Zephyr OS build v4.1.0-5079-g6caa1017b0ce ***
*** Booting Zephyr OS build v4.1.0-5079-g6caa1017b0ce ***
Address of sample 0x70000000
Hello sysbuild with mcuboot! nucleo_h7s3l8
```

titouanc commented:
Hello @FRASTM and thank you for following that up.

In just pulled your latest changes from a few minutes ago, and I still encounter the same issue of the application not starting up on my STM32H7S78-DK.

With that being said, I built for the same board as you did (nucleo_h7s3l8), and then I compared the generated `zephyr.dts` to spot any key difference. From there, I noticed that on the STM32H7S78-DK, there's a PSRAM on XSPI1 that is not present on the Nucleo board.
To test further, I removed the entire xspi1 node from the board overlay here https://github.com/zephyrproject-rtos/zephyr/blob/0ca62307b64814a1ba08096d04c13d7c10b207b8/boards/st/stm32h7s78_dk/stm32h7s78_dk.dts#L197-L222 and after that... :drum: it works !

```
*** Booting MCUboot v2.1.0-rc1-315-ge3b8bd449b18 ***
*** Using Zephyr OS build v4.1.0-5079-ga78196b40245 ***
*** Booting Zephyr OS build v4.1.0-5079-ga78196b40245 ***
Address of sample 0x70000000
Hello sysbuild with mcuboot! stm32h7s78_dk
```

I will try to understand why XSPI1 (PSRAM) is interfering with XSPI2 (FLASH), but let me know if you have any hint.

etienne-lms reviewed (approved):
LGTM (FWIW)
